DVCodingStyle.md

```diff
@@ -0,0 +1,864 @@
+    should be used for every `uvm_env` and `uvm_agent` component in the testbench.
+3.  It is recommended that the environment's configuration object contain the
+    agent's configuration object.
+4.  The `uvm_config_db` API should be used to pass configuration objects to
+    locations where they are needed. It should not be used to pass integers,
+    strings, or other basic data types, since it is much easier for namespace
+    collisions to occur when using lower level types.
+5.  You must always check for success when accessing the `uvm_config_db`.
+    If the lookup fails, issue a `uvm_fatal` to terminate the simulation.
+    This should be done as follows:
+
+    ```systemverilog
+    if (!uvm_config_db#(...)::get(...)) begin
+      `uvm_fatal(...)
+    end
+    ```
+6.  Wildcards must not be used in the field argument of `uvm_config_db::set(...)` calls.
+    It is allowed to use wildcards in the inst_name argument of these calls,
+    with proper precautions.
+
+
+### Sequences
+
+1.  It is recommended to keep sequences as generic as possible and to avoid
+    writing directed sequences unless absolutely necessary.
+2.  It is recommended to avoid explicit delays in terms of absolute time to
+    make code emulation friendly. Delays in terms of clock cycles are allowed.
+3.  The `body()` method should only execute the raw functional behavior of
+    the sequence.
+    Any associated housekeeping code should be placed elsewhere, such as in the
+    `pre_start()` and `post_start()` methods.
+    *   `pre_start()` and `post_start()` must be used instead of `pre_body()`
+        and `post_body()`, as these will always be called, even for subsequences
+        called from a parent sequence.
+4.  When creating a sequence object, always randomize the sequence before starting it.
+5.  Virtual sequences must be used to coordinate the behavior of multiple agents and multiple sequences.
+6.  Virtual sequences must be started on the null sequencer or on a valid virtual sequencer.
+
+
+### Sequence Items
+
+It is recommended, but not necessary, to create small "unit tests" for
+transactions objects to ensure the randomization constraints are working as
+intended.
+
+
+### RAL Usage
+
+When using the UVM RAL model, care must be taken when using `uvm_reg::set()` and
+`uvm_reg::update()`, as these are both non-atomic operations that can easily
+cause race conditions if there are any parallel threads.
+
+
+### Objections and Coordinating End of Test
+
+1.  At a high level, a phase should not end when either there is more stimulus
+    to send during that phase or the DUT has yet to respond to some stimulus that
+    has been sent.
+    To prevent a phase from ending in the former case, raise and lower objections
+    inside tests.
+    To prevent a phase from ending in the latter case, raise objections in the
+    scoreboard whenever expect queues are not empty and lower them when they are empty.
```

Comment:
I completely disagree with the comments in line 602 og 603.
I would say to avoid using objections in scoreboards and use them in monitors instead.

Raising objections in scoreboard can lead to deadlocks.  Ideally objections should only be in the test. 
but if needed a monitor can raise and objection in the Phase_ready_to_end phase if data is seen.

Also raising objections outside of a test should only be done in the phase_ready_to_end to improve performance.

maybe the topic of objections warrants separate thread?

Reply:
I see some valid usecases in allowing raising/dropping objections in the scoreboard; example: in xbar type DUT, raise obj when transaction enters the DUT and drop only when it exits. This objection (that helps ensure that the test does not exit prematurely) cannot be captured in the monitor and has to be captured in a higher level component. It cannot be captured in the test either since it is an entity that is only generating and bombarding C-R stimulus into the DUT. It is no knowledge of when the transactions enter and exit the DUT. 

We keep the performance impact low by raising objection exactly once (dont raise if it is already raised), which is what the line implies. You can find that logic in our base scoreboard (`hw/dv/sv/dv_lib/dv_base_scoreboard.sv`). If by deadlocks you mean hangs, then there is an obvious problem either in the design or dv, that needs fixing. We avoid hanging forever through reasonable test timeouts.

I have not used `phase_ready_to_end` and would like to learn more about it. Let's discuss this in a separate thread.

Reply:
I have created #30 do discuss further